### ISSUE TITLE
Benchmark speed: 1.26 lower then NTSC05

### ISSUE THREAD

**Issue by tomcw** [AppleWin/AppleWin#424]
Spun off from #261, where Michael commented "(Video) Benchmark speeds lower then NTSC05".
NB. May also want to consider #401.

Testing on my desktop PC (Win7-64, AMD Phenom II dual-core @3ghz), for these versions/configurations:
- NTSC05 : NTSC Color Monitor
- 1.26.2.4 : Color Monitor
- 1.25.0.4 : Color (Standard)

I get this data:

AppleWin | Pure Video FPS (hires) | Pure Video FPS (text) | Pure CPU MHz | Expected perf (FPS)
--|--|--|--|--
NTSC05 | 334 | 333 | 17.3 | 335
1.26.2.4 | 405 | 403 | 12.4 | 376
1.25.0.4 | 398 | 400 | 203.5 | 405

So it's the "Pure CPU MHz" number that we may want to understand.

NB. The AppleWin config UI allows a max emulation speed of 3.9 MHz and beyond that it's unthrottled. So this "Pure CPU MHz" is only useful for the unthrottled case.

**Comment by tomcw** [comment]
In `VideoBenchmark()`, the benchmark for "Pure CPU MHz" is calling `CpuExecute()` with bVideoUpdate=true. But when AppleWin is running unthrottled, then bVideoUpdate=false.

Re-running with `CpuExecute(100000, /*bVideoUpdate=/*false)`, then we get:
```
Pure CPU MHz:	227.8
```

This is faster than 1.25.0.4, which could be explained by the 16% speed-up in #347 (except here it's only a 12% speed up: 203.5 -> 227.8). It's "ball-park" at least.

Still this doesn't explain the slowdown between NTSC05 and 1.26.2.4.

**Comment by Michaelangel007** [comment]
@tomcw Hey Tom, is there a specific commit for NTSC05 ? Thanks.

**Comment by tomcw** [comment]
There's a steady degradation in _Pure CPU MHz_ from 1.26.2.4 to 1.26.3.6:

AppleWin | Pure Video FPS (hires) | Pure Video FPS (text) | Pure CPU MHz | Expected perf (FPS)
--|--|--|--|--
1.26.2.4 | 2860 | 2819 | 12.4 | 574
1.26.3.0 | 2864 | 2830 | 12.4 | 572
1.26.3.1 | 2857 | 2834 | 12.2 | 568
1.26.3.2 | 2871 | 2840 | 12.0 | 551
1.26.3.4 | 2869 | 2872 | 11.6 | 547
1.26.3.5 | 2860 | 2861 | 11.6 | 548
1.26.3.6 | 2867 | 2834 | 11.7 | 546

Again tested on my desktop PC (Win7-64, AMD Phenom II dual-core @3ghz).

**Comment by tomcw** [comment]
AppleWin | Pure Video FPS (hires) | Pure Video FPS (text) | Pure CPU MHz (video update) | Pure CPU MHz (full-speed) | Expected perf (FPS)
--|--|--|--|--|--
1.26.3.6 | 398 | 406 | 12.8 | n/a | 375
1.27.1.0 | 398 | 400 | 13.1 | 236.1 | 377
1.27.2.0 | 398 | 406 | 13.1 | 237.9 | 374
1.27.4.0 | 405 | 403 | 13.0 | 238.1 | 377
1.27.5.0 | 405 | 406 | 12.9 | 236.0 | 376
1.27.6.0 | 398 | 406 | 13.1 | 237.6 | 373
1.27.7.0 | 405 | 402 | 12.8 | 234.9 | 374
1.27.8.0 | 398 | 400 | 12.9 | 239.2 | 377
1.27.9.0 | 402 | 399 | 12.8 | 242.2 | 377
1.27.10.0 | 398 | 399 | 13.0 | 241.9 | 376
1.27.11.0 | 405 | 406 | 12.8 | 242.8 | 377
1.27.12.0 | 405 | 399 | 12.6 | 240.8 | 376

Config:
- Enhanced //e.
- No cards in slot 4, 5, 7.
- Video mode = 50% Amber Monitor.

I don't know why these "Pure Video" numbers vary so much from January. I re-ran v1.26.3.6 as a control, and you can see it's consistent with the new numbers now!

**Comment by tomcw** [comment]
I don't think so, since the pre-Spectre numbers (May'17, ie. the first comment in this issue) match today's numbers (Nov'18).

And...
>AMD said it has already released microcode updates to its customers and partners for all chips dating back to the first Bulldozer core products released in 2011.

My Phenom II is the predecessor to Bulldozer ([ref](https://en.wikipedia.org/wiki/AMD_10h)).

But you could be right that it's a Windows driver/update related thing.
btw. this Phenom II based machine is still Win7-64.

**Comment by tomcw** [comment]
A look at how 'Color (RGB Monitor)' performs:

Config:
- AppleWin 1.28.0.0
- Enhanced //e.
- No cards in slot 4, 5, 7.
- AMD Phenom II based machine on Win7-64.

### Video mode = 50% Amber Monitor (x1 Windowed)

AppleWin | Pure Video FPS (hires) | Pure Video FPS (text) | Pure CPU MHz (video update) | Pure CPU MHz (full-speed) | Expected perf (FPS)
--|--|--|--|--|--
1.28.0.0 | 398 | 406 | 12.9 | 230.2 | 568
1.28.0.0 | 401 | 400 | 12.7 | 230.2 | 374
1.28.0.0 | 405 | 407 | 12.4 | 229.0 | 451
**mean** | 401.3 | 404.3 | 12.7 | 229.8 | 464.3

### Video mode = 50% Color (RGB Monitor) (x1 Windowed)

AppleWin | Pure Video FPS (hires) | Pure Video FPS (text) | Pure CPU MHz (video update) | Pure CPU MHz (full-speed) | Expected perf (FPS)
--|--|--|--|--|--
1.28.0.0 | 398 | 403 | 17.1 | 227.7 | 505
1.28.0.0 | 405 | 407 | 16.9 | 233.0 | 384
1.28.0.0 | 398 | 400 | 15.9 | 221.7 | 588
**mean** | 400.3 | 403.3 | 16.6 | 227.5 | 492.3

So a mean +4MHz (or +33%) speed up for the "Pure CPU MHz (video update)". This is running both CPU emulation and Windows video updates every opcode. This mode only exists in this benchmark, and not in regular emulation. It gives an idea of the maximum non-full-speed MHz speed (eg. for the Config dialog's emulation speed slider).

"Pure CPU MHz (full-speed)" as expected doesn't change, this works by only updating the Windows display every wall-clock 60Hz - ie. the video rendering speed is much less important here since it's only done very occasionally.

**Comment by tomcw** [comment]
A look at how 'Color (RGB Monitor)' with **vertical blend** performs (#616):

Config:
- AppleWin 1.28.5.0
- Enhanced //e.
- No cards in slot 4, 5, 7.
- AMD Phenom II based machine on Win7-64.

### Video mode = 50% Color (RGB Monitor) (x1 Windowed), _no vertical blend_

AppleWin | Pure Video FPS (hires) | Pure Video FPS (text) | Pure CPU MHz (video update) | Pure CPU MHz (full-speed) | Expected perf (FPS)
--|--|--|--|--|--
1.28.5.0 | 405 | 407 | 17.1 | 233.1 | 848
1.28.5.0 | 398 | 400 | 17.2 | 234.6 | 510
1.28.5.0 | 400 | 400 | 17.1 | 233.8 | 609
**mean** | 401.0 | 402.3 | 17.1 | 233.8 | 655.7

### Video mode = 50% Color (RGB Monitor) (x1 Windowed), _vertical blend_

AppleWin | Pure Video FPS (hires) | Pure Video FPS (text) | Pure CPU MHz (video update) | Pure CPU MHz (full-speed) | Expected perf (FPS)
--|--|--|--|--|--
1.28.5.0 | 404 | 397 | 7.6 | 230.5 | 327
1.28.5.0 | 405 | 402 | 7.6 | 228.4 | 328
1.28.5.0 | 398 | 402 | 7.5 | 225.9 | 330
**mean** | 402.3 | 400.3 | 7.6 | 228.3 | 328.3

So with **vertical blending on**, the "Pure CPU MHz (video update)" is 55% slower.
